File: services/event-gateway/internal/events/envelope.go
package events

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"net"
	"net/http"
	"strings"
	"time"

	"github.com/google/uuid"
)

const Version = "1.0"

type Category string

const (
	Identity       Category = "identity"
	Session        Category = "session"
	Media          Category = "media"
	Playback       Category = "playback"
	Creator        Category = "creator"
	Community      Category = "community"
	Recommendation Category = "recommendation"
	Search         Category = "search"
	Moderation     Category = "moderation"
	Commerce       Category = "commerce"
	System         Category = "system"
)

type Metadata struct {
	Platform      string  `json:"platform"`
	Region        string  `json:"region"`
	IPHash        *string `json:"ip_hash"`
	UserAgentHash *string `json:"user_agent_hash"`
}

type Envelope struct {
	EventID        string         `json:"event_id"`
	EventType      string         `json:"event_type"`
	EventVersion   string         `json:"event_version"`
	EventCategory  Category       `json:"event_category"`
	Producer       string         `json:"producer"`
	Timestamp      string         `json:"timestamp"`
	UserID         *string        `json:"user_id"`
	CreatorID      *string        `json:"creator_id"`
	MediaID        *string        `json:"media_id"`
	SessionID      *string        `json:"session_id"`
	DeviceID       *string        `json:"device_id"`
	RequestID      *string        `json:"request_id"`
	CorrelationID  *string        `json:"correlation_id"`
	IdempotencyKey string         `json:"idempotency_key"`
	Payload        map[string]any `json:"payload"`
	Metadata       Metadata       `json:"metadata"`
}

var (
	ErrUnknownCategory = errors.New("unknown event category")
	ErrUnknownType     = errors.New("unknown event type")
)

var allowedTypes = map[Category]map[string]struct{}{
	Identity:       set("user_created", "profile_updated", "user_deleted"),
	Session:        set("session_started", "session_ended", "device_registered"),
	Media:          set("upload_started", "upload_completed", "media_processing_started", "media_processing_completed", "media_published"),
	Playback:       set("playback_started", "playback_paused", "playback_resumed", "playback_completed", "playback_buffered", "playback_seeked", "playback_error"),
	Creator:        set("creator_profile_created", "creator_verified", "creator_content_published"),
	Community:      set("community_created", "community_member_joined"),
	Recommendation: set("feed_requested", "recommendation_served", "recommendation_clicked", "recommendation_skipped"),
	Search:         set("search_performed", "search_result_clicked"),
	Moderation:     set("content_flagged", "moderation_action_taken"),
	Commerce:       set("checkout_started", "purchase_completed"),
	System:         set("service_started", "service_error", "background_job_completed"),
}

var topicByCategory = map[Category]string{
	Identity:       "identity.events.v1",
	Session:        "session.events.v1",
	Media:          "media.events.v1",
	Playback:       "playback.events.v1",
	Creator:        "creator.events.v1",
	Community:      "community.events.v1",
	Recommendation: "recommendation.events.v1",
	Search:         "search.events.v1",
	Moderation:     "moderation.events.v1",
	Commerce:       "commerce.events.v1",
	System:         "system.events.v1",
}

func Decode(raw []byte) (Envelope, error) {
	dec := json.NewDecoder(strings.NewReader(string(raw)))
	dec.DisallowUnknownFields()
	var envelope Envelope
	if err := dec.Decode(&envelope); err != nil {
		return Envelope{}, fmt.Errorf("decode event envelope: %w", err)
	}
	if err := envelope.Validate(); err != nil {
		return Envelope{}, err
	}
	return envelope, nil
}

func (e Envelope) Validate() error {
	if _, err := uuid.Parse(e.EventID); err != nil {
		return fmt.Errorf("event_id must be uuid: %w", err)
	}
	if e.EventVersion != Version {
		return fmt.Errorf("event_version must be %s", Version)
	}
	if _, ok := allowedTypes[e.EventCategory]; !ok {
		return ErrUnknownCategory
	}
	if _, ok := allowedTypes[e.EventCategory][e.EventType]; !ok {
		return fmt.Errorf("%w %q for category %q", ErrUnknownType, e.EventType, e.EventCategory)
	}
	if strings.TrimSpace(e.Producer) == "" {
		return errors.New("producer is required")
	}
	if _, err := time.Parse(time.RFC3339, e.Timestamp); err != nil {
		return fmt.Errorf("timestamp must be ISO-8601/RFC3339: %w", err)
	}
	for field, value := range map[string]*string{
		"user_id":    e.UserID,
		"creator_id": e.CreatorID,
		"session_id": e.SessionID,
	} {
		if value != nil {
			if _, err := uuid.Parse(*value); err != nil {
				return fmt.Errorf("%s must be uuid when present: %w", field, err)
			}
		}
	}
	if strings.TrimSpace(e.IdempotencyKey) == "" {
		return errors.New("idempotency_key is required")
	}
	if e.Metadata.Platform != "web" && e.Metadata.Platform != "mobile" && e.Metadata.Platform != "tv" && e.Metadata.Platform != "service" {
		return errors.New("metadata.platform must be web, mobile, tv, or service")
	}
	if strings.TrimSpace(e.Metadata.Region) == "" {
		return errors.New("metadata.region is required")
	}
	if e.Payload == nil {
		return errors.New("payload is required")
	}
	if err := validatePayloadShape(e.EventType, e.Payload); err != nil {
		return err
	}
	if err := rejectPrivateFields(e.Payload, "payload"); err != nil {
		return err
	}
	return nil
}

func (e Envelope) Topic() string {
	return topicByCategory[e.EventCategory]
}

func (e Envelope) PartitionKey() string {
	for _, value := range []*string{e.MediaID, e.CreatorID, e.UserID, e.SessionID, e.DeviceID, e.CorrelationID} {
		if value != nil && *value != "" {
			return *value
		}
	}
	return e.EventID
}

func (e Envelope) WithRequestMetadata(r *http.Request) Envelope {
	region := r.Header.Get("X-NEXT-Region")
	if e.Metadata.Region == "" && region != "" {
		e.Metadata.Region = region
	}
	if e.Metadata.IPHash == nil {
		hash := hashRemoteIP(r.RemoteAddr)
		e.Metadata.IPHash = &hash
	}
	if e.Metadata.UserAgentHash == nil {
		hash := hashString(r.UserAgent())
		e.Metadata.UserAgentHash = &hash
	}
	return e
}

func (e Envelope) Marshal() ([]byte, error) {
	return json.Marshal(e)
}

func DLQTopic(category Category) string {
	switch category {
	case Identity:
		return "identity.events.dlq.v1"
	case Media:
		return "media.events.dlq.v1"
	case Playback:
		return "playback.events.dlq.v1"
	default:
		return "analytics.events.dlq.v1"
	}
}

func set(values ...string) map[string]struct{} {
	result := make(map[string]struct{}, len(values))
	for _, value := range values {
		result[value] = struct{}{}
	}
	return result
}

func rejectPrivateFields(value any, path string) error {
	switch typed := value.(type) {
	case map[string]any:
		for key, child := range typed {
			lower := strings.ToLower(key)
			switch lower {
			case "ip", "ip_address", "raw_ip", "precise_location", "latitude", "longitude", "password", "token", "access_token", "refresh_token", "private_message", "payment_secret", "full_user_agent", "user_agent":
				return fmt.Errorf("privacy violation: raw private field %s.%s is not allowed", path, key)
			}
			if err := rejectPrivateFields(child, path+"."+key); err != nil {
				return err
			}
		}
	case []any:
		for i, child := range typed {
			if err := rejectPrivateFields(child, fmt.Sprintf("%s[%d]", path, i)); err != nil {
				return err
			}
		}
	}
	return nil
}

var requiredPayloadFields = map[string][]string{
	"user_created":               {"consent_analytics"},
	"profile_updated":            {"changed_fields", "profile_version"},
	"user_deleted":               {"deletion_reason", "anonymized"},
	"session_started":            {"auth_method"},
	"session_ended":              {"duration_ms", "end_reason"},
	"device_registered":          {"device_family", "os_family"},
	"upload_started":             {"upload_id", "file_size_bytes", "mime_type"},
	"upload_completed":           {"upload_id", "file_size_bytes", "duration_ms", "checksum_sha256"},
	"media_processing_started":   {"pipeline_id", "pipeline_version"},
	"media_processing_completed": {"pipeline_id", "duration_ms", "renditions"},
	"media_published":            {"visibility"},
	"playback_started":           {"position_ms"},
	"playback_paused":            {"position_ms"},
	"playback_resumed":           {"position_ms"},
	"playback_completed":         {"duration_ms", "watch_time_ms", "completion_ratio"},
	"playback_buffered":          {"position_ms", "buffer_duration_ms"},
	"playback_seeked":            {"from_ms", "to_ms"},
	"playback_error":             {"error_code", "fatal"},
	"creator_profile_created":    {"creator_handle"},
	"creator_verified":           {"verification_level", "verified_by"},
	"creator_content_published":  {"content_type", "visibility"},
	"community_created":          {"community_id", "visibility"},
	"community_member_joined":    {"community_id"},
	"feed_requested":             {"surface", "request_reason", "limit"},
	"recommendation_served":      {"surface", "algorithm", "candidate_count", "served_media_ids"},
	"recommendation_clicked":     {"surface", "rank", "algorithm"},
	"recommendation_skipped":     {"surface", "rank", "dwell_ms"},
	"search_performed":           {"query_hash", "query_length", "result_count"},
	"search_result_clicked":      {"query_hash", "result_rank", "result_type"},
	"content_flagged":            {"content_type", "reason_code", "source"},
	"moderation_action_taken":    {"action", "reason_code", "policy_version"},
	"checkout_started":           {"checkout_id", "currency", "amount_minor"},
	"purchase_completed":         {"checkout_id", "currency", "amount_minor", "payment_reference_hash"},
	"service_started":            {"service_name", "service_version", "environment"},
	"service_error":              {"service_name", "error_code", "severity", "retryable"},
	"background_job_completed":   {"job_name", "duration_ms", "status"},
}

func validatePayloadShape(eventType string, payload map[string]any) error {
	for _, field := range requiredPayloadFields[eventType] {
		if _, ok := payload[field]; !ok {
			return fmt.Errorf("payload.%s is required for %s", field, eventType)
		}
	}
	for _, field := range []string{"position_ms", "duration_ms", "watch_time_ms", "buffer_duration_ms", "from_ms", "to_ms", "rank", "dwell_ms", "candidate_count", "limit", "result_count", "query_length", "file_size_bytes", "amount_minor"} {
		if value, ok := payload[field]; ok {
			number, ok := value.(float64)
			if !ok || number < 0 {
				return fmt.Errorf("payload.%s must be a non-negative number", field)
			}
		}
	}
	return nil
}

func hashRemoteIP(remoteAddr string) string {
	host, _, err := net.SplitHostPort(remoteAddr)
	if err != nil {
		host = remoteAddr
	}
	return hashString(host)
}

func hashString(value string) string {
	sum := sha256.Sum256([]byte(value))
	return hex.EncodeToString(sum[:])
}

File: packages/events/src/contracts/envelope.ts
import { randomUUID } from 'node:crypto';
import { z } from 'zod';

export const eventCategories = [
  'identity',
  'session',
  'media',
  'playback',
  'creator',
  'community',
  'recommendation',
  'search',
  'moderation',
  'commerce',
  'system',
] as const;

export const eventPlatforms = ['web', 'mobile', 'tv', 'service'] as const;

export type EventCategory = (typeof eventCategories)[number];
export type EventPlatform = (typeof eventPlatforms)[number];

const nullableUuid = z.string().uuid().nullable();
const nullableHash = z.string().min(16).max(128).nullable();
const stringId = z.string().min(1).max(128);
const nullableStringId = stringId.nullable();
const nonNegativeInt = z.number().int().nonnegative();

export const eventMetadataSchema = z
  .object({
    platform: z.enum(eventPlatforms),
    region: z.string().min(2).max(64),
    ip_hash: nullableHash,
    user_agent_hash: nullableHash,
  })
  .strict();

export type EventMetadata = z.infer<typeof eventMetadataSchema>;

const identityPayloads = {
  user_created: z
    .object({
      handle: z.string().min(3).max(30).optional(),
      email_hash: nullableHash.optional(),
      signup_source: z.string().max(64).optional(),
      consent_analytics: z.boolean(),
    })
    .strict(),
  profile_updated: z
    .object({
      changed_fields: z.array(z.string().min(1).max(64)).min(1),
      profile_version: z.number().int().positive(),
    })
    .strict(),
  user_deleted: z
    .object({
      deletion_reason: z.enum([
        'user_requested',
        'policy_action',
        'legal_request',
        'retention_expired',
      ]),
      anonymized: z.literal(true),
    })
    .strict(),
} as const;

const sessionPayloads = {
  session_started: z
    .object({
      auth_method: z.enum(['password', 'passkey', 'oauth', 'magic_link', 'service']),
      client_version: z.string().max(64).optional(),
    })
    .strict(),
  session_ended: z
    .object({
      duration_ms: nonNegativeInt,
      end_reason: z.enum(['logout', 'expired', 'revoked', 'client_closed', 'unknown']),
    })
    .strict(),
  device_registered: z
    .object({
      device_family: z.enum(['desktop', 'phone', 'tablet', 'tv', 'service']),
      os_family: z.string().min(1).max(64),
      push_enabled: z.boolean().optional(),
    })
    .strict(),
} as const;

const mediaPayloads = {
  upload_started: z
    .object({
      upload_id: stringId,
      file_size_bytes: nonNegativeInt,
      mime_type: z.string().min(3).max(128),
    })
    .strict(),
  upload_completed: z
    .object({
      upload_id: stringId,
      file_size_bytes: nonNegativeInt,
      duration_ms: nonNegativeInt,
      checksum_sha256: z.string().length(64),
    })
    .strict(),
  media_processing_started: z
    .object({
      pipeline_id: stringId,
      pipeline_version: z.string().min(1).max(64),
    })
    .strict(),
  media_processing_completed: z
    .object({
      pipeline_id: stringId,
      duration_ms: nonNegativeInt,
      renditions: z.array(z.string().min(1).max(32)).min(1),
    })
    .strict(),
  media_published: z
    .object({
      visibility: z.enum(['public', 'followers', 'unlisted', 'private']),
      tags: z.array(z.string().min(1).max(64)).max(32).default([]),
      language: z.string().min(2).max(16).optional(),
    })
    .strict(),
} as const;

const playbackPayloads = {
  playback_started: z
    .object({
      position_ms: nonNegativeInt,
      rendition: z.string().min(1).max(32).optional(),
      startup_latency_ms: nonNegativeInt.optional(),
    })
    .strict(),
  playback_paused: z.object({ position_ms: nonNegativeInt }).strict(),
  playback_resumed: z.object({ position_ms: nonNegativeInt }).strict(),
  playback_completed: z
    .object({
      duration_ms: nonNegativeInt,
      watch_time_ms: nonNegativeInt,
      completion_ratio: z.number().min(0).max(1),
    })
    .strict(),
  playback_buffered: z
    .object({
      position_ms: nonNegativeInt,
      buffer_duration_ms: nonNegativeInt,
      rendition: z.string().min(1).max(32).optional(),
    })
    .strict(),
  playback_seeked: z
    .object({
      from_ms: nonNegativeInt,
      to_ms: nonNegativeInt,
    })
    .strict(),
  playback_error: z
    .object({
      error_code: z.string().min(1).max(64),
      fatal: z.boolean(),
      position_ms: nonNegativeInt.optional(),
    })
    .strict(),
} as const;

const creatorPayloads = {
  creator_profile_created: z
    .object({
      creator_handle: z.string().min(3).max(30),
      category: z.string().min(1).max(64).optional(),
    })
    .strict(),
  creator_verified: z
    .object({
      verification_level: z.enum(['identity', 'organization', 'notable']),
      verified_by: z.string().min(1).max(128),
    })
    .strict(),
  creator_content_published: z
    .object({
      content_type: z.enum(['video', 'live', 'post', 'course', 'product']),
      visibility: z.enum(['public', 'followers', 'unlisted', 'private']),
    })
    .strict(),
} as const;

const communityPayloads = {
  community_created: z
    .object({
      community_id: z.string().uuid(),
      visibility: z.enum(['public', 'private', 'invite_only']),
    })
    .strict(),
  community_member_joined: z
    .object({
      community_id: z.string().uuid(),
      join_source: z.string().min(1).max(64).optional(),
    })
    .strict(),
} as const;

const recommendationPayloads = {
  feed_requested: z
    .object({
      surface: z.enum(['home', 'explore', 'creator', 'search', 'related']),
      request_reason: z.enum(['initial_load', 'pagination', 'refresh', 'preload']),
      limit: z.number().int().positive().max(200),
    })
    .strict(),
  recommendation_served: z
    .object({
      surface: z.enum(['home', 'explore', 'creator', 'search', 'related']),
      algorithm: z.string().min(1).max(128),
      candidate_count: nonNegativeInt,
      served_media_ids: z.array(stringId).max(200),
      experiment_ids: z.array(z.string().min(1).max(64)).max(32).default([]),
    })
    .strict(),
  recommendation_clicked: z
    .object({
      surface: z.enum(['home', 'explore', 'creator', 'search', 'related']),
      rank: nonNegativeInt,
      algorithm: z.string().min(1).max(128),
    })
    .strict(),
  recommendation_skipped: z
    .object({
      surface: z.enum(['home', 'explore', 'creator', 'search', 'related']),
      rank: nonNegativeInt,
      dwell_ms: nonNegativeInt,
    })
    .strict(),
} as const;

const searchPayloads = {
  search_performed: z
    .object({
      query_hash: nullableHash,
      query_length: z.number().int().min(0).max(512),
      filters: z.record(z.string(), z.string()).default({}),
      result_count: nonNegativeInt,
    })
    .strict(),
  search_result_clicked: z
    .object({
      query_hash: nullableHash,
      result_rank: nonNegativeInt,
      result_type: z.enum(['media', 'creator', 'community', 'commerce', 'learn']),
    })
    .strict(),
} as const;

const moderationPayloads = {
  content_flagged: z
    .object({
      content_type: z.enum(['media', 'profile', 'comment', 'live', 'community']),
      reason_code: z.string().min(1).max(64),
      source: z.enum(['user_report', 'automated_model', 'trusted_flagger', 'staff']),
    })
    .strict(),
  moderation_action_taken: z
    .object({
      action: z.enum(['no_action', 'age_gate', 'label', 'limit_distribution', 'remove', 'suspend']),
      reason_code: z.string().min(1).max(64),
      policy_version: z.string().min(1).max(64),
    })
    .strict(),
} as const;

const commercePayloads = {
  checkout_started: z
    .object({
      checkout_id: stringId,
      currency: z.string().length(3),
      amount_minor: nonNegativeInt,
    })
    .strict(),
  purchase_completed: z
    .object({
      checkout_id: stringId,
      currency: z.string().length(3),
      amount_minor: nonNegativeInt,
      payment_reference_hash: nullableHash,
    })
    .strict(),
} as const;

const systemPayloads = {
  service_started: z
    .object({
      service_name: z.string().min(1).max(128),
      service_version: z.string().min(1).max(64),
      environment: z.enum(['dev', 'staging', 'prod', 'test']),
    })
    .strict(),
  service_error: z
    .object({
      service_name: z.string().min(1).max(128),
      error_code: z.string().min(1).max(64),
      severity: z.enum(['debug', 'info', 'warn', 'error', 'critical']),
      retryable: z.boolean(),
    })
    .strict(),
  background_job_completed: z
    .object({
      job_name: z.string().min(1).max(128),
      duration_ms: nonNegativeInt,
      status: z.enum(['success', 'failed', 'cancelled']),
    })
    .strict(),
} as const;

export const eventPayloadSchemas = {
  ...identityPayloads,
  ...sessionPayloads,
  ...mediaPayloads,
  ...playbackPayloads,
  ...creatorPayloads,
  ...communityPayloads,
  ...recommendationPayloads,
  ...searchPayloads,
  ...moderationPayloads,
  ...commercePayloads,
  ...systemPayloads,
} as const;

export type EventType = keyof typeof eventPayloadSchemas;

export const eventCategoryByType = {
  user_created: 'identity',
  profile_updated: 'identity',
  user_deleted: 'identity',
  session_started: 'session',
  session_ended: 'session',
  device_registered: 'session',
  upload_started: 'media',
  upload_completed: 'media',
  media_processing_started: 'media',
  media_processing_completed: 'media',
  media_published: 'media',
  playback_started: 'playback',
  playback_paused: 'playback',
  playback_resumed: 'playback',
  playback_completed: 'playback',
  playback_buffered: 'playback',
  playback_seeked: 'playback',
  playback_error: 'playback',
  creator_profile_created: 'creator',
  creator_verified: 'creator',
  creator_content_published: 'creator',
  community_created: 'community',
  community_member_joined: 'community',
  feed_requested: 'recommendation',
  recommendation_served: 'recommendation',
  recommendation_clicked: 'recommendation',
  recommendation_skipped: 'recommendation',
  search_performed: 'search',
  search_result_clicked: 'search',
  content_flagged: 'moderation',
  moderation_action_taken: 'moderation',
  checkout_started: 'commerce',
  purchase_completed: 'commerce',
  service_started: 'system',
  service_error: 'system',
  background_job_completed: 'system',
} as const satisfies Record<EventType, EventCategory>;

export const eventTypesByCategory = eventCategories.reduce(
  (acc, category) => {
    acc[category] = Object.entries(eventCategoryByType)
      .filter(([, value]) => value === category)
      .map(([key]) => key as EventType);
    return acc;
  },
  {} as Record<EventCategory, EventType[]>,
);

export type EventPayload<TEventType extends EventType> = z.infer<
  (typeof eventPayloadSchemas)[TEventType]
>;

const baseEventEnvelopeSchema = z.object({
  event_id: z.string().uuid(),
  event_type: z.string().min(1),
  event_version: z.literal('1.0'),
  event_category: z.enum(eventCategories),
  producer: z.string().min(1).max(128),
  timestamp: z.string().datetime({ offset: true }),
  user_id: nullableUuid,
  creator_id: nullableUuid,
  media_id: nullableStringId,
  session_id: nullableUuid,
  device_id: nullableStringId,
  request_id: z.string().max(256).nullable(),
  correlation_id: z.string().max(256).nullable(),
  idempotency_key: z.string().min(8).max(256),
  payload: z.record(z.string(), z.unknown()),
  metadata: eventMetadataSchema,
});

export const eventEnvelopeSchema = baseEventEnvelopeSchema.superRefine((event, ctx) => {
  if (!isEventType(event.event_type)) {
    ctx.addIssue({
      code: z.ZodIssueCode.custom,
      path: ['event_type'],
      message: `Unknown event type: ${event.event_type}`,
    });
    return;
  }

  const expectedCategory = eventCategoryByType[event.event_type];
  if (event.event_category !== expectedCategory) {
    ctx.addIssue({
      code: z.ZodIssueCode.custom,
      path: ['event_category'],
      message: `Event ${event.event_type} belongs to ${expectedCategory}`,
    });
  }

  const payloadResult = eventPayloadSchemas[event.event_type].safeParse(event.payload);
  if (!payloadResult.success) {
    for (const issue of payloadResult.error.issues) {
      ctx.addIssue({
        ...issue,
        path: ['payload', ...issue.path],
      });
    }
  }
});

export type EventEnvelope = z.infer<typeof eventEnvelopeSchema>;

export function isEventType(value: string): value is EventType {
  return Object.prototype.hasOwnProperty.call(eventPayloadSchemas, value);
}

export function validateEventEnvelope(input: unknown): EventEnvelope {
  return eventEnvelopeSchema.parse(input);
}

export function safeValidateEventEnvelope(
  input: unknown,
): z.SafeParseReturnType<unknown, EventEnvelope> {
  return eventEnvelopeSchema.safeParse(input);
}

export interface CreateEventEnvelopeInput<TEventType extends EventType> {
  readonly event_type: TEventType;
  readonly producer: string;
  readonly payload: EventPayload<TEventType>;
  readonly metadata: EventMetadata;
  readonly event_id?: string;
  readonly timestamp?: string;
  readonly user_id?: string | null;
  readonly creator_id?: string | null;
  readonly media_id?: string | null;
  readonly session_id?: string | null;
  readonly device_id?: string | null;
  readonly request_id?: string | null;
  readonly correlation_id?: string | null;
  readonly idempotency_key?: string;
}

export function createEventEnvelope<TEventType extends EventType>(
  input: CreateEventEnvelopeInput<TEventType>,
): EventEnvelope {
  const eventId = input.event_id ?? randomUUID();
  const envelope = {
    event_id: eventId,
    event_type: input.event_type,
    event_version: '1.0',
    event_category: eventCategoryByType[input.event_type],
    producer: input.producer,
    timestamp: input.timestamp ?? new Date().toISOString(),
    user_id: input.user_id ?? null,
    creator_id: input.creator_id ?? null,
    media_id: input.media_id ?? null,
    session_id: input.session_id ?? null,
    device_id: input.device_id ?? null,
    request_id: input.request_id ?? null,
    correlation_id: input.correlation_id ?? null,
    idempotency_key: input.idempotency_key ?? `${input.producer}:${input.event_type}:${eventId}`,
    payload: input.payload,
    metadata: input.metadata,
  };
  return validateEventEnvelope(envelope);
}

export function partitionKeyForEvent(event: EventEnvelope): string {
  return (
    event.media_id ??
    event.creator_id ??
    event.user_id ??
    event.session_id ??
    event.device_id ??
    event.correlation_id ??
    event.event_id
  );
}
